### ISSUE TITLE
Reduction of code footprint

### ISSUE THREAD

**Issue by sigitas-stumbras** [abap2xlsx/abap2xlsx#524]
This is more of a feature suggestion to reduce the code footprint when using abap2xlsx, not an actual issue.
1. lo_style_boxed = lo_excel->add_new_style( **io_clone_of = lo_style_base** ).
2. lo_excel_worksheet->**set_cell( )** to have an optional **io_style** paramenter, to pass reference to zcl_excel_style directly instead of requiring additional variable top pass guid to ip_style parameter.
3. zcl_excel_worksheet->**set_cell( )** and **change_cell_style( )** to have an optional **ip_cell** paramater, that would accept Excel style cell adress. Examples:
lo_excel_worksheet->set_cell( **ip_cell = 'B1'** io_style = lo_style ip_value = 'value' ).
lo_excel_worksheet->set_cell( **ip_cell = 'C1:E3'** io_style = lo_style ip_value = 'value' ). - Merger and all affected cells are handled automatically.
**ip_row_to** and **ip_column_to** parameters could also be included in both methods as an alternative capability to specify a range of cells

I could do the programing, but I would like to hear some feedback first. I really find this project quite amazing and usefull and I would like to contribute.

**Comment by ivanfemia** [comment]
Hi,

1. I like the idea could be useful
1. maybe we can consider to remove the guid functionality at all (in the future? to guarantee backward compatibility). we may need to do an impact analysis on this.
1. Cool functionality

Thank you for your availability to contribute!
Ivan

**Comment by sigitas-stumbras** [comment]
Hi,

I had some second thoughts considering the second point in my suggestions.

Although adding additional optional io_style parameter is a trivial matter, I suspect you may be concerned over not overextending the interface of the method. Especially considering this in long term backward compatibility.

Ok. Here is an alternative suggestion - how about making ip_style a generic type parameter, that would accept both guid and object reference. So the same parameter could be handled with some dynamic programing, without adding additional parameter and remaining fully backward-compatible.

Sigitas

A side note: Please be open and straightforward, if you have any comments or I am doing something not quite right github-wise, contribution-wise or otherwise. I have never done this before, but I'd really like to get some first hand experience with this. Especialy - to get the hang of how these kind of projects get managed and project decisions being made.

**Comment by sigitas-stumbras** [comment]
Hello again,

Some second thoughts regarding point 3.
After getting to understand the project and it's codebase a bit better I now realize that adding this Excel-style cell reference parameter **ip_cell** just to the two methods is a bit ad-hoc. To be entirely consistent, a whole range of getters and setters should be accepting this parameter. Should I go forward with this?

**Comment by ivanfemia** [comment]
I would prefer a consistent change that a quick and dirty fix!
Thank you a lot for your contribution
Ivan

**Comment by sandraros** [comment]
@sigitas-stumbras Sorry to be so late on checking your proposal. Please tell me your opinion if we change it a little bit like that:
- I propose to not implement the range (e.g. `C1:E3`) via `set_cell` and `change_cell_style` because it's the job of `set_area` and `set_area_style`, but instead to add `ip_range` in `set_area` and `set_area_style` (notice that I prefer `ip_range` over `ip_cell` here).
- About adding `ip_cell` and range processing in the `change_cell_style` (which works for "static styles"), I would prefer to reuse a method like `set_area_style` and make it call `change_cell_style` in a loop, but I'm reluctant to do so because its 107 parameters work on the principle of `IS SUPPLIED`. I would prefer to first do a PR to propose a new class to set the static styles of one cell, and then it would be easier to add the static styles in `set_area_style` (and refactor `change_cell_style`).
  - Refactoring of `change_cell_style` done (#904)
- Concerning the improvement to pass the style as an instance instead of passing its GUID (parameter `ip_style` in methods `set_cell` (EDIT adding this: <ins>making it of type `any` so that to accept either GUID or `type ref to zcl_excel_style`</ins>), etc.), I fully agree (EDIT deleting this: <del>that it's much more convenient, but I would prefer to add a new parameter `io_style`</del>).
- Not sure if we should rewrite all the demo programs in the same PR or in another PR so that to keep the main PR small.

Thank you.
@AndreaBorgia-Abo fyi and opinion too

**Comment by AndreaBorgia-Abo** [comment]
@sandraros Fully agree, 107 parameters is a monster to be tamed first.
With respect to the contents to the future PR, I'd prefer having two separate PRs to review, staged in time, than a single jumbo PR: just reference the same bug, without tagging it "fix" so that GH won't close the report prematurely.

@sigitas-stumbras asking for feedback is already a good sign that you have the right attitude, I like that.
Especially because participating in this project is a (re)learning experience for me as well!
On the topic of getters and setters, I've [recently learned they're evil :)](https://blogs.sap.com/2021/10/07/public-read-only-the-best-invention-since-sliced-bread-or-an-unspeakable-travesty/#comment-593327)

**Comment by sandraros** [comment]
I realize that `SET_AREA` and `SET_AREA_FORMULA` don't apply a value or a formula on the area transmitted (I think #574 was merged without checking it thoroughly). I created #867. There's also another related bug #866.
- Change was done (#869)

**Comment by sandraros** [comment]
@AndreaBorgia-Abo I have proposed the new PR #930. I ended with the parameter names `IP_COLUMNROW` (one cell) and `IP_RANGE` (multiple cells) instead of `IP_CELL`. NB: the first one is named after parameter `I_COLUMNROW` which is used by `ZCL_EXCEL_COMMON=>CONVERT_COLUMNROW2COLUMN_A_ROW`.

List of methods changed in `ZCL_EXCEL_WORKSHEET` and the demonstration programs which have been modified:
- `CHANGE_AREA_STYLE`
  - Demonstration in program `zdemo_excel40`
- `CHANGE_CELL_STYLE`
  - Demonstration in program `zdemo_excel35`
- `GET_CELL`
  - Demonstration in program `zdemo_excel15`
- `SET_AREA`
  - Demonstration in program `zdemo_excel40`
- `SET_AREA_HYPERLINK`
  - No demonstration program
- `SET_AREA_STYLE`
  - No demonstration program
- `SET_CELL`
  - Demonstration in programs `zdemo_excel1`, `zdemo_excel13` and `zdemo_excel35`
- `SET_CELL_FORMULA`
  - No demonstration program
- `SET_CELL_STYLE`
  - Demonstration in program `zdemo_excel35`
- `SET_MERGE`
  - Demonstration in program `zdemo_excel13`
- `SET_MERGE_STYLE`
  - No demonstration program
- `NORMALIZE_COLUMNROW_PARAMETER` (created)
  - test class added to test this private method
- `NORMALIZE_RANGE_PARAMETER` (created)
  - test class added to test this private method

`ZCL_EXCEL_COMMON`:
- `CONVERT_COLUMNROW2COLUMN_A_ROW`
  - New parameter `E_COLUMN_INT`
  - test class refactored and completed for new parameter
- `CONVERT_RANGE2COLUMN_A_ROW`
  - New parameters `E_COLUMN_START_INT` and `E_COLUMN_END_INT`.
  - test class refactored and completed for new parameters

I used this test program for all methods above, especially `SET_AREA_HYPERLINK`, `SET_AREA_STYLE`, `SET_CELL_FORMULA` and `SET_MERGE_STYLE`:
```abap
REPORT.
TABLES sscrfields.
DATA: gc_save_file_name TYPE string VALUE 'test.xlsx'.
INCLUDE zdemo_excel_outputopt_incl.
DATA: lo_excel TYPE REF TO zcl_excel.

AT SELECTION-SCREEN.
  CHECK sscrfields-ucomm = 'ONLI'.
  DATA: lx_root TYPE REF TO cx_root.
  TRY.
      lo_excel = NEW zcl_excel( ).
      DATA(lo_worksheet) = lo_excel->get_active_worksheet( ).

      "==================
      " SET_AREA
      "==================
      lo_worksheet->set_area( ip_range = 'A1:B2' ip_formula = '"new R"&ROW()&"C"&COLUMN()' ip_area = lo_worksheet->c_area-whole ).
      lo_worksheet->set_area( ip_column_start = 'D' ip_column_end = 'E' ip_row = 1 ip_row_to = 2 ip_formula = '"old R"&ROW()&"C"&COLUMN()' ip_area = lo_worksheet->c_area-whole ).

      "==================
      " CHANGE_AREA_STYLE
      "==================
      lo_worksheet->CHANGE_AREA_STYLE( ip_range = 'A1:B1'
                                       ip_style_changer = zcl_excel_style_changer=>create( lo_excel
                                                          )->set_fill_filltype( zcl_excel_style_fill=>c_fill_solid
                                                          )->set_fill_fgcolor_rgb( zcl_excel_style_color=>c_yellow ) ).
      lo_worksheet->CHANGE_AREA_STYLE( ip_column_start = 'A' ip_row = 1 ip_row_to = 2
                                       ip_style_changer = zcl_excel_style_changer=>create( lo_excel
                                                          )->set_font_color( value #( rgb = zcl_excel_style_color=>c_red ) ) ).

      "==================
      " CHANGE_CELL_STYLE
      "==================
      lo_worksheet->CHANGE_cell_STYLE( ip_columnrow = 'D1'
                                       ip_fill_filltype = zcl_excel_style_fill=>c_fill_solid
                                       ip_fill_fgcolor_rgb = zcl_excel_style_color=>c_yellow ).
      lo_worksheet->CHANGE_cell_STYLE( ip_column = 'D' ip_row = 1
                                       ip_font_color_rgb = zcl_excel_style_color=>c_red ).

      "==================
      " GET_CELL
      "==================
      lo_worksheet->get_cell( EXPORTING ip_columnrow = 'B3'
                              IMPORTING ep_value = DATA(value) ).
      lo_worksheet->get_cell( EXPORTING ip_column = 'B' ip_row = 3
                              IMPORTING ep_value = DATA(value2) ).
      lo_worksheet->get_cell( EXPORTING ip_column = 2 ip_row = 3
                              IMPORTING ep_value = DATA(value3) ).
      TRY.
          lo_worksheet->get_cell( EXPORTING ip_columnrow = 'B3' ip_column = 2 ip_row = 3
                                  IMPORTING ep_value = DATA(value4) ).
          MESSAGE 'get_cell too many row/column parameters' TYPE 'E'.
        CATCH cx_root INTO lx_root.
      ENDTRY.
      TRY.
          lo_worksheet->get_cell( IMPORTING ep_value = DATA(value5) ).
          MESSAGE 'get_cell missing row/column parameter' TYPE 'E'.
        CATCH cx_root INTO lx_root.
      ENDTRY.

      "==================
      " SET_AREA_HYPERLINK
      "==================
      lo_worksheet->set_area_hyperlink( ip_range = 'D1:E1' ip_url = 'https://www.google.com' ip_is_internal = abap_false ).
      lo_worksheet->set_area_hyperlink( ip_column_start = 'D' ip_column_end = 'E' ip_row = 2 ip_url = 'https://www.sap.com' ip_is_internal = abap_false ).

      "==================
      " SET_AREA_STYLE
      "==================
      DATA(style_font_blue) = lo_excel->add_new_style( ).
      style_font_blue->font->color-rgb = zcl_excel_style_color=>c_blue.
      lo_worksheet->set_area_style( ip_range = 'D1:E1' ip_style = style_font_blue->get_guid( ) ).
      lo_worksheet->set_area_style( ip_column_start = 'D' ip_column_end = 'E' ip_row = 2 ip_style = style_font_blue->get_guid( ) ).

      "==================
      " SET_CELL
      "==================
      lo_worksheet->set_cell( ip_columnrow = 'B3' ip_value = CONV decfloat16( '0.3-' ) ).
      lo_worksheet->set_cell( ip_column = 'C' ip_row = 4 ip_value = 'text' ).
      lo_worksheet->set_cell( ip_column = 4 ip_row = 5 ip_value = 'text' ).
      TRY.
          lo_worksheet->set_cell( ip_value = 'text' ).
          MESSAGE 'set_cell missing row/column parameter' TYPE 'E'.
        CATCH cx_root INTO lx_root.
      ENDTRY.
      TRY.
          lo_worksheet->set_cell( ip_columnrow = 'B3' ip_column = 'B' ip_row = 3 ip_value = 'text' ).
          MESSAGE 'set_cell too many row/column parameters' TYPE 'E'.
        CATCH cx_root INTO lx_root.
      ENDTRY.

      "==================
      " SET_CELL_FORMULA
      "==================
      lo_worksheet->set_cell_formula( ip_columnrow = 'C3' ip_formula = '2*-0.15' ).
      lo_worksheet->set_cell_formula( ip_column = 'A' ip_row = 4 ip_formula = '"te"&"xt"' ).

      "==================
      " SET_CELL_STYLE
      "==================
      lo_worksheet->set_cell_style( ip_columnrow = 'B3' ip_style = style_font_blue->get_guid( ) ).
      lo_worksheet->set_cell_style( ip_column = 'C' ip_row = 3 ip_style = style_font_blue->get_guid( ) ).

      "==================
      " SET_MERGE
      "==================
      DATA(style_center) = lo_excel->add_new_style( ).
      style_center->alignment->horizontal = zcl_excel_style_alignment=>c_horizontal_center.
      style_center->alignment->vertical = zcl_excel_style_alignment=>c_vertical_center.
      lo_worksheet->set_merge( ip_range = 'C3:D3' ip_style = style_center->get_guid( ) ).
      lo_worksheet->set_merge( ip_column_start = 'B' ip_row = 3 ip_row_to = 4 ip_style = style_center->get_guid( ) ).

      "==================
      " SET_MERGE_STYLE
      "==================
      DATA(style_fill_yellow) = lo_excel->add_new_style( ).
      style_fill_yellow->fill->filltype = zcl_excel_style_fill=>c_fill_solid.
      style_fill_yellow->fill->fgcolor-rgb = zcl_excel_style_color=>c_yellow.
      lo_worksheet->set_merge_style( ip_range = 'A4:A5' ip_style = style_fill_yellow->get_guid( ) ).
      lo_worksheet->set_merge_style( ip_column_start = 'D' ip_column_end = 'E' ip_row = 5 ip_row_to = 6 ip_style = style_fill_yellow->get_guid( ) ).


    CATCH cx_root INTO lx_root.
      MESSAGE lx_root TYPE 'E'.
  ENDTRY.
  ASSERT 1 = 1. " debug helper

START-OF-SELECTION.
  lcl_output=>output( cl_excel = lo_excel ).
  ASSERT 1 = 1. " debug helper
```

**Comment by AndreaBorgia-Abo** [comment]
@sandraros should I wait until after #940 is merged before reviewing this one? To help you decide, I'll add that in a couple of days I'll have some spare time.

**Comment by sandraros** [comment]
Better adopt #940 sooner, but I let you choose 😉.

**Comment by AndreaBorgia-Abo** [comment]
#930 is approved, merge at your convenience.

**Comment by sandraros** [comment]
I'd like to integrate #929 also, it's currently draft, I have to review it, and after that, we're done with this issue.

**Comment by AndreaBorgia-Abo** [comment]
Would you like a pre-review? Otherwise, I'll wait until it is finalized.

**Comment by sandraros** [comment]
#929 is now ready for review.

**Comment by sandraros** [comment]
All pull requests merged, closing.
